src/functional/ffront/mockitir.py

```diff
@@ -0,0 +1,215 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from typing import Union
+
+from functional.iterator import ir as itir
+
+
+def sym(sym_or_name: Union[str, itir.Sym]) -> itir.Sym:
+    """
+    Convert to Sym if necessary.
+
+    Examples
+    --------
+    >>> sym("a")
+    Sym(id='a')
+
+    >>> sym(itir.Sym(id="b"))
+    Sym(id='b')
+    """
+    if isinstance(sym_or_name, itir.Sym):
+        return sym_or_name
+    return itir.Sym(id=sym_or_name)
+
+
+def ref(ref_or_name: Union[str, itir.SymRef]) -> itir.SymRef:
+    """
+    Convert to SymRef if necessary.
+
+    Examples
+    --------
+    >>> ref("a")
+    SymRef(id='a')
+
+    >>> ref(itir.SymRef(id="b"))
+    SymRef(id='b')
+    """
+    if isinstance(ref_or_name, itir.SymRef):
+        return ref_or_name
+    return itir.SymRef(id=ref_or_name)
+
+
+def exprify(literal_or_expr: Union[str, int, itir.Expr]) -> itir.Expr:
+    """
+    Convert into a SymRef or IntLiteral if necessary.
+
+    Examples
+    --------
+    >>> exprify("a")
+    SymRef(id='a')
+
+    >>> exprify(3)
+    IntLiteral(value=3)
+
+    >>> exprify(itir.OffsetLiteral(value="i"))
+    OffsetLiteral(value='i')
+    """
+    if isinstance(literal_or_expr, str):
+        return ref(literal_or_expr)
+    elif isinstance(literal_or_expr, int):
+        return itir.IntLiteral(value=literal_or_expr)
+    return literal_or_expr
+
+
+class lambda_:
+    """
+    Create a lambda from params and an expression.
+
+    Examples
+    --------
+    >>> lambda_("a")(deref("a"))  # doctest: +ELLIPSIS
```

Comment:
Not sure I'm a fan of this staggered creation. I don't really see the need to do that for a lambda.

Reply:
I am convinced, if you spend some days building complex enough snippets, you will appreciate it as much as I do.

Reply:
I don't mind changing it if you still think it's better after considering the following:

They are both not an exact match.
```python
(lambda a, b, c: expr)(1, 2, 3)
call(lambda_(a, b, c)(expr))(1, 2, 3)
call(lambda_([a, b, c], expr))(1, 2, 3)
```
However the former captures that python syntax is not `(lambda [a, b, c], expr)(1, 2, 3)` for good reason. The `:` helps visually.

What it gives you when you actually use it is

```python
call(
    lambda(*vars)(
        (
            super_long_expression
            ...
        )
    )
)(*values)
```
as opposed to
```python
call(
    lambda(
        vars, 
        (
            super_long_expression
            ...
        )
    )
)(*values)
```

Which may not seem like much in such a simple example, but note how the `vars` line up with the values they are given above, and how they don't below. This slight advantage is multiplied by the complexity of the snippet.

Reply:
Just to clarify, I'm fine with both the original proposal and Till's proposal, so for me this is not a blocker. Iif you don't manage to agree here, I'd then suggest to go with the original proposal since it's already implemented and it was explicitly designed that way.